### PR TITLE
New rule for complex expression

### LINKED ISSUE #241: Describe rules for calls chain inside parenthesis
In line like `if (node.treeParent.findChildByType(MODIFIER_LIST)?.findChildByType(OVERRIDE_KEYWORD) != null)` is it better to simply follow functional style
```kotlin
if (node
    .treeParent
    .findChildByType(MODIFIER_LIST)
    ?.findChildByType(OVERRIDE_KEYWORD) != null)
```
or suggest to extract complex expression into separate variable?

### LINKED ISSUE #676: Long dot qualified expression in condition should be replaced with new variable
In this case:
```kotlin
if (node
    .treeParent
    .findChildByType(MODIFIER_LIST)
    ?.findChildByType(OVERRIDE_KEYWORD) != null)
```
you should replace with new variable to: 
```kotlin
val a = node
    .treeParent
    .findChildByType(MODIFIER_LIST)
    ?.findChildByType(OVERRIDE_KEYWORD)
if (a != null)
```

### PR DESCRIPTION
**Pull request by kentr0w**
### What's done:
Created new rule

## Which rule and warnings did you add?

This pull request closes 

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [x] Updated diktat-analysis.yml
* [x] Updated available-rules.md

closes #676 
closes #241

### CODE REVIEW

File: diktat-rules/src/main/resources/diktat-analysis-huawei.yml [hunk 1]
```diff
@@ -145,6 +145,8 @@
   enabled: true
   configuration:
     maxParametersInOneLine: 2
+- name: COMPLEX_EXPRESSION
+  enabled: true
```
Comment (aktsay6) [inline diktat-rules/src/main/resources/diktat-analysis-huawei.yml:149]: Please provide some comments for rule in each yml file

**Review by aktsay6 — CHANGES_REQUESTED** [review]

File: info/available-rules.md [hunk 2]
```diff
@@ -67,6 +67,7 @@
...
 |    3  |     3.6.2  |  REDUNDANT_SEMICOLON                      |  Check: warns if semicolons are used at the end of line.<br>Fix: removes semicolon.                                                                                            |  yes  |  no  |                                                                                      - |
 |    3  |     3.6.2  |  WRONG_NEWLINES                           |  Check: warns if line breaks do not follow code style guid.<br>Fix: fixes incorrect line breaks.                                                                               |  yes  |  no  |                                                                                      - |
+|    3  |     3.6.2  |  COMPLEX_EXPRESSION                       |  Check: warns if long dot expression using in condition or as arguments                                                                                                        |  no   |  no  | - |
```
Comment (petertrr) [inline info/available-rules.md:70]:
```suggestion
|    3  |     3.6.2  |  COMPLEX_EXPRESSION                       |  Check: warns if a long dot qualified expression is used in condition or as an argument                                                                                                         |  no   |  no  | - |
```

File: info/guide/guide-chapter-3.md [hunk 3]
```diff
@@ -366,6 +366,19 @@ Note that all comparison operators, such as `==`, `>`, `<`, should not be split.
 ```kotlin
 if (condition) list.map { foo(it) }.filter { bar(it) } else list.drop(1)
 ```  
+
+**Note:** If dot qualified expression is inside condition or passed as an argument - replace with new vatiable
```
Comment (petertrr) [inline info/guide/guide-chapter-3.md:370]:
```suggestion
**Note:** If dot qualified expression is inside condition or passed as an argument - replace with new variable
```

**Review by petertrr — CHANGES_REQUESTED** [review]
Please check snapshot diktat check, there are a lot of new error there.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/AnnotationNewLineRule.kt [hunk 4]
```diff
@@ -68,7 +68,12 @@ class AnnotationNewLineRule(private val configRules: List<RulesConfig>) : Rule("
             if (node == node.treeParent.getFirstChildWithType(node.elementType)) {
                 // Current node is ANNOTATION_ENTRY. treeParent(ModifierList) -> treeParent(PRIMARY_CONSTRUCTOR)
                 // Checks if there is a white space before grandparent node
-                if (node.treeParent.treeParent.treePrev.isWhiteSpace()) {
+                val isParentWhiteSpace = node
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/AnnotationNewLineRule.kt:71]: `hasSpaceBeforeGrandparent` or something similar?

File: info/guide/guide-chapter-3.md [hunk 5]
```diff
@@ -366,6 +366,19 @@ Note that all comparison operators, such as `==`, `>`, `<`, should not be split.
 ```kotlin
 if (condition) list.map { foo(it) }.filter { bar(it) } else list.drop(1)
 ```  
+
+**Note:** If dot qualified expression is inside condition or passed as an argument - replace with new variable
```
Comment (petertrr) [inline info/guide/guide-chapter-3.md:370]:
```suggestion
**Note:** If dot qualified expression is inside condition or passed as an argument - it should be replaced with new variable.
```

File: info/guide/guide-chapter-3.md [hunk 6]
```diff
@@ -366,6 +366,19 @@ Note that all comparison operators, such as `==`, `>`, `<`, should not be split.
...
+
+**Invalid example**: 
+```kotlin
+     if (node.text.length.dec() != 0) {}
```
Comment (petertrr) [inline info/guide/guide-chapter-3.md:374]: This is a strange example, why would you call `dec` here?

File: info/guide/guide-chapter-3.md [hunk 7]
```diff
@@ -366,6 +366,19 @@ Note that all comparison operators, such as `==`, `>`, `<`, should not be split.
...
+```
+ 
+**Valid example**: 
+```kotlin
+        val nodeLen = node.text.length.dec()
```
Comment (petertrr) [inline info/guide/guide-chapter-3.md:379]: This is not a very good variable name to be used in an example :) It's not node length, it's node's *text* length minus one. Please either find a more straightforward example or rename the variable to be more self-descriptive.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 8]
```diff
@@ -504,6 +515,15 @@ class NewlinesRule(private val configRules: List<RulesConfig>) : Rule("newlines"
             firstChildNode.elementType == IDENTIFIER &&
             treeParent.elementType == BINARY_EXPRESSION
 
+    /**
+     * This method checks that complex expression should be replace with new variable
+     */
+    private fun ASTNode.isInParentheses() = parent({it.elementType == DOT_QUALIFIED_EXPRESSION || it.elementType == SAFE_ACCESS_EXPRESSION})
+        ?.treeParent
+        ?.elementType
+        ?.let { it in bracketsTypes }
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:524]: `parenthesesTypes`?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 9]
```diff
@@ -439,17 +444,23 @@ class NewlinesRule(private val configRules: List<RulesConfig>) : Rule("newlines"
...
      * @return true - if there is error, and false if there is no error
      */
-    private fun ASTNode.isCallsChain() = getCallChain()?.isNotValidCalls(this) ?: false
-
-    private fun ASTNode.getCallChain() = getParentExpressions()
-        .lastOrNull()
-        ?.run {
-            mutableListOf<ASTNode>().also {
-                getOrderedCallExpressions(psi, it)
+    private fun ASTNode.isCallsChain(isWithoutParentheses: Boolean = true) = getCallChain(isWithoutParentheses)?.isNotValidCalls(this) ?: false
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:447]: What is the meaning of `isWithoutParentheses` parameter? If it is intended to get all calls without dropping first ones, which we don't count for functional-style-check, it should be called differently.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 10]
```diff
@@ -375,11 +380,11 @@ class NewlinesRule(private val configRules: List<RulesConfig>) : Rule("newlines"
         if (psi.children.isNotEmpty() && (!psi.isFirstChildElementType(DOT_QUALIFIED_EXPRESSION) &&
                 !psi.isFirstChildElementType(SAFE_ACCESS_EXPRESSION))) {
             val firstChild = psi.firstChild
+            if (firstChild.isFirstChildElementType(DOT_QUALIFIED_EXPRESSION) ||
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:383]: Why is this block moved?

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt (hunk 10 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:383]: Because there was incorrect code

Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:386]: In what way? Why was it working before? If it wasn't, was in covered in incorrect tests? Have you updated those tests?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 11]
```diff
@@ -444,17 +447,17 @@ class NewlinesRule(private val configRules: List<RulesConfig>) : Rule("newlines"
      *
      * @return true - if there is error, and false if there is no error
      */
-    private fun ASTNode.isCallsChain(isWithoutParentheses: Boolean = true) = getCallChain(isWithoutParentheses)?.isNotValidCalls(this) ?: false
+    private fun ASTNode.isCallsChain(dropLeadingBrackets: Boolean = true) = getCallChain(dropLeadingBrackets)?.isNotValidCalls(this) ?: false
 
-    private fun ASTNode.getCallChain(isWithoutParentheses: Boolean = true): List<ASTNode>? {
+    private fun ASTNode.getCallChain(dropLeadingBrackets: Boolean = true): List<ASTNode>? {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:452]: We are not dropping brackets, we are dropping parts of dot-qualified expression which are not function calls.

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt (hunk 10 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:386]: There was an `if` inside another `if` that could never execute, I changed, and there is a test covering new `if`.

**Review by petertrr — APPROVED** [review]
LGTM

**Review by aktsay6 — APPROVED** [review]
lgtm

**Pull request merged**